File: packages/typescript/src/_package_paths.ts
/**
 * Classify filesystem paths as ghagen-internal vs. user-authored.
 *
 * Shared by `_source_location.ts` (stack-frame attribution for model
 * factories) and `pin/sources.ts` (deps-upgrade user-file tracking) so a
 * single prefix computation and predicate serve both.
 *
 * `import.meta.url` always points at this file. Its directory is the
 * "package source root" — `.../packages/typescript/src/` in development or
 * `.../node_modules/@ghagen/ghagen/dist/` once published. Any path under
 * that prefix is treated as internal, EXCEPT ghagen's own test files (paths
 * containing `.test.`), so that ghagen's tests can attribute themselves.
 */

const PACKAGE_INTERNAL_DIR: string = (() => {
  try {
    const url = new URL(".", import.meta.url);
    const path = url.pathname;
    return path.endsWith("/") ? path : `${path}/`;
  } catch {
    return "";
  }
})();

/**
 * True if `filename` belongs to ghagen's own source/dist or to
 * `node_modules` — i.e. NOT a user-authored frame. Ghagen's own `.test.`
 * files are treated as external so their frames can be captured.
 */
export function isInternalFrame(filename: string): boolean {
  if (filename.includes("/node_modules/")) {
    return true;
  }
  if (PACKAGE_INTERNAL_DIR && filename.startsWith(PACKAGE_INTERNAL_DIR)) {
    // Allow `.test.` files through so ghagen's own unit tests can capture
    // themselves as the call site.
    if (filename.includes(".test.")) {
      return false;
    }
    return true;
  }
  return false;
}

/**
 * True if `path` is a real user-authored source file: a filesystem path
 * that is neither ghagen-internal, under `node_modules`, nor a non-file
 * URL (`data:` / `node:`). This is the tracking-side counterpart to
 * {@link isInternalFrame}, with the extra URL/empty filtering that
 * `jiti.cache` keys can require.
 */
export function isUserFile(path: string): boolean {
  if (!path) {
    return false;
  }
  // Skip URLs that aren't real filesystem paths.
  if (path.startsWith("data:") || path.startsWith("node:")) {
    return false;
  }
  return !isInternalFrame(path);
}

File: packages/python/src/ghagen/_package_paths.py
"""Classify filesystem paths as ghagen-internal vs. user-authored.

Shared by :mod:`ghagen.models._base` (stack-frame attribution for model
factories) and :mod:`ghagen.pin.sources` (deps-upgrade user-file tracking) so a
single ghagen-root computation and predicate serve both — the Python peer of
TypeScript's ``_package_paths.ts``.

The two callers ask the question with different remits, so this module exposes
*two* predicates over one shared root rather than forcing one algorithm on both:

* :func:`is_internal_frame` — stack attribution. A frame is internal when it is
  inside pydantic or the ghagen package; the first non-internal frame is the
  user's call site.
* :func:`is_user_file` — import tracking. A file is user-authored when it is
  neither under ``site-packages``, the ghagen package, nor the stdlib.
"""

from __future__ import annotations

import sysconfig
from pathlib import Path

#: Root of the installed ``ghagen`` package (…/ghagen). ``__file__`` is
#: ``…/ghagen/_package_paths.py``, so its parent is the package directory. Any
#: frame or file inside it is ghagen-internal, regardless of the submodule
#: (models/, emitter/, helpers/, pin/, …).
GHAGEN_ROOT = Path(__file__).resolve().parent


def is_internal_frame(filename: str) -> bool:
    """Return True if ``filename`` is inside pydantic or the ghagen package.

    Used by stack-frame attribution to skip framework frames when locating the
    user's model-construction call site.
    """
    if "pydantic" in Path(filename).parts:
        return True
    try:
        resolved = Path(filename).resolve()
    except (OSError, ValueError):  # pragma: no cover — defensive
        return False
    return resolved == GHAGEN_ROOT or GHAGEN_ROOT in resolved.parents


def is_user_file(path: Path) -> bool:
    """Return True if *path* is a user-authored file (not stdlib/site/ghagen).

    Used by deps-upgrade file tracking to keep the tracked set to files the
    user controls.
    """
    parts = path.parts

    # Exclude site-packages
    if "site-packages" in parts:
        return False

    # Exclude ghagen package files
    try:
        path.relative_to(GHAGEN_ROOT)
        return False
    except ValueError:
        pass

    # Exclude stdlib files
    stdlib_paths = sysconfig.get_paths()
    for key in ("stdlib", "platstdlib", "purelib", "platlib"):
        stdlib_dir = stdlib_paths.get(key)
        if stdlib_dir:
            try:
                path.relative_to(stdlib_dir)
                return False
            except ValueError:
                pass

    return True
